In [0]:
from __future__ import absolute_import, division, print_function, unicode_literals
import tensorflow as tf
from keras import optimizers


Load and prepare the [MNIST](http://yann.lecun.com/exdb/mnist/) dataset. Convert the samples from integers to floating-point numbers:

In [0]:
mnist = tf.keras.datasets.mnist

(x_train, y_train), (x_test, y_test) = mnist.load_data()
x_train, x_test = x_train / 255.0, x_test / 255.0

## Model with 1 hidden layer

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


sgd = tf.keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(optimizer=sgd,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

Train and evaluate model:

In [26]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 6s 108us/sample - loss: 0.5180 - acc: 0.8559
Epoch 2/5
60000/60000 [==============================] - 6s 106us/sample - loss: 0.3173 - acc: 0.9080
Epoch 3/5
60000/60000 [==============================] - 6s 106us/sample - loss: 0.2908 - acc: 0.9160
Epoch 4/5
60000/60000 [==============================] - 6s 105us/sample - loss: 0.2693 - acc: 0.9220
Epoch 5/5
10000/10000 [==============================] - 1s 97us/sample - loss: 0.2391 - acc: 0.9320


[0.23908916844427586, 0.932]

## Model with 3 hidden layer

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


sgd = tf.keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(optimizer=sgd,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [28]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 9s 145us/sample - loss: 1.3876 - acc: 0.5167
Epoch 2/5
60000/60000 [==============================] - 9s 142us/sample - loss: 0.4227 - acc: 0.8760
Epoch 3/5
60000/60000 [==============================] - 9s 142us/sample - loss: 0.3476 - acc: 0.8993
Epoch 4/5
60000/60000 [==============================] - 9s 143us/sample - loss: 0.3092 - acc: 0.9108
Epoch 5/5
10000/10000 [==============================] - 1s 106us/sample - loss: 0.2536 - acc: 0.9261


[0.25361136218309405, 0.9261]

## Model with 9 hidden layer

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


sgd = tf.keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(optimizer=sgd,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [30]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 15s 244us/sample - loss: 2.3297 - acc: 0.1024
Epoch 2/5
60000/60000 [==============================] - 14s 234us/sample - loss: 2.3066 - acc: 0.1061
Epoch 3/5
60000/60000 [==============================] - 14s 235us/sample - loss: 2.3034 - acc: 0.1080
Epoch 4/5
60000/60000 [==============================] - 14s 238us/sample - loss: 2.3031 - acc: 0.1081
Epoch 5/5
10000/10000 [==============================] - 1s 135us/sample - loss: 2.3021 - acc: 0.1135


[2.3020533473968507, 0.1135]

In [0]:
model = tf.keras.models.Sequential([
  tf.keras.layers.Flatten(input_shape=(28, 28)),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(512, activation='sigmoid'),
  tf.keras.layers.BatchNormalization(axis=-1, epsilon=0.001, center=True, scale=True, beta_initializer='zeros', gamma_initializer='ones', moving_mean_initializer='zeros', moving_variance_initializer='ones', beta_regularizer=None, gamma_regularizer=None, beta_constraint=None, gamma_constraint=None),
  tf.keras.layers.Dense(10, activation=tf.nn.softmax)
])


sgd = tf.keras.optimizers.SGD(lr=0.01, decay=1e-6, momentum=0.9, nesterov=True)
model.compile(optimizer=sgd,
              loss='sparse_categorical_crossentropy',
              metrics=['accuracy'])

In [32]:
model.fit(x_train, y_train, epochs=5)

model.evaluate(x_test, y_test)

Epoch 1/5
60000/60000 [==============================] - 39s 651us/sample - loss: 0.3969 - acc: 0.8867
Epoch 2/5
60000/60000 [==============================] - 38s 628us/sample - loss: 0.2304 - acc: 0.9329
Epoch 3/5
60000/60000 [==============================] - 38s 628us/sample - loss: 0.1801 - acc: 0.9468
Epoch 4/5
60000/60000 [==============================] - 38s 640us/sample - loss: 0.1535 - acc: 0.9546
Epoch 5/5
10000/10000 [==============================] - 2s 191us/sample - loss: 0.0920 - acc: 0.9738


[0.0919776431108825, 0.9738]

Hier sieht man deutlich, dass durch BN zwischen allen Layern das Ergebnis deutlich gestiegen ist. **Von knapp 11% auf 97%.** 